compact/range.go

```diff
@@ -111,6 +110,29 @@ func (r *Range) AppendRange(other *Range, visitor VisitFn) error {
 	return r.appendImpl(other.end, other.hashes[0], other.hashes[1:], visitor)
 }
 
+// Merge extends the compact range by merging in the other compact range from
+// the right. It uses the tree hasher to calculate hashes of newly created
+// nodes, and reports them through the visitor function (if non-nil). The other
+// range must begin between the current range's begin and end.
+//
+// Warning: This method modifies both this and the other Range.
+// Warning: This method is experimental.
+func (r *Range) Merge(other *Range, visitor VisitFn) error {
+	if _, err := r.intersectImpl(other); err != nil {
+		return err
+	}
+	return r.AppendRange(other, visitor)
+}
+
+// Intersect returns the intersection of two compact ranges. The other range
+// must begin between the current range's begin and end.
+//
+// Warning: This method modifies both this and the other Range.
+// Warning: This method is experimental.
+func (r *Range) Intersect(other *Range) (*Range, error) {
```

Comment:
The modification of both ranges feels especially confusing here.  Really after you call this function you have three compact ranges:
- `r` is now `r` without the intersecting parts.
- the output is the intersection.
- `other` is now `other` without the intersecting parts.
This seems like it could lead to even more confusion than the behavior of `Merge` above.  Surely what you'd want is just the intersection, like the function name suggests, without all of the side effects?